Windows: daemon fails to start on Java 21 — Unable to establish loopback connection (AF_UNIX / WEPollSelectorImpl)

## Environment
- OS: Windows 11 Pro (10.0.26200)
- Java: Azul Zulu 21.0.10 (build 21.0.10+7-LTS)
- kcp-memory: v0.22.0
- Running inside: Claude Code Desktop (MSIX packaged app)

## Error

The daemon fails immediately on startup with:

```
java.io.IOException: Unable to establish loopback connection
    at sun.nio.ch.PipeImpl$Initializer.run(PipeImpl.java:103)
    at sun.nio.ch.WEPollSelectorImpl.<init>(WEPollSelectorImpl.java:79)
    at sun.nio.ch.WEPollSelectorProvider.openSelector(WEPollSelectorProvider.java:33)
    at java.nio.channels.Selector.open(Selector.java:295)
    at com.sun.net.httpserver.HttpServer.create(HttpServer.java:152)
Caused by: java.net.SocketException: Invalid argument: connect
    at sun.nio.ch.UnixDomainSockets.connect0(Native Method)
    at sun.nio.ch.PipeImpl$Initializer$LoopbackConnector.run(PipeImpl.java:132)
```

## Root Cause

Java 21's `WEPollSelectorImpl` uses AF_UNIX sockets internally for its NIO pipe implementation. When running inside the Claude Code MSIX sandbox, AF_UNIX socket creation fails even though `afunix.sys` is present on the system. This matches OpenJDK bug [JDK-8312215](https://bugs.openjdk.org/browse/JDK-8312215).

`sc query afunix` returns nothing (driver not loaded as a service), confirming AF_UNIX is unavailable in this context.

## Attempted Workarounds (all failed)
- `-Djava.nio.channels.spi.SelectorProvider=sun.nio.ch.WindowsSelectorProvider`
- `-Djava.net.preferIPv4Stack=true`
- `-Djava.io.tmpdir=C:\Temp` (simpler path)
- `--enable-native-access=ALL-UNNAMED`
- Java 17 — rejected: JAR compiled for class file version 65.0 (Java 21)

## Workaround That Works

Running the daemon **outside** the MSIX sandbox (from a regular PowerShell terminal) works fine:

```powershell
& "C:\Program Files\Zulu\zulu-21\bin\java.exe" -jar "$env:USERPROFILE\.kcp\kcp-memory-daemon.jar" daemon
```

Port 7735 is then reachable from within Claude Code. Auto-start configured via Windows Startup folder bat file.

## Suggested Fix

Add a JVM flag or launcher option to force TCP-based NIO pipes instead of AF_UNIX, or document the MSIX sandbox limitation with the PowerShell workaround. A dedicated `kcp-memory.bat` Windows launcher shipped with the installer would also help.

--- a/java/src/main/java/com/cantara/kcp/memory/KcpMemoryDaemon.java
+++ b/java/src/main/java/com/cantara/kcp/memory/KcpMemoryDaemon.java
@@ -5,11 +5,10 @@ import com.cantara.kcp.memory.mcp.McpServer;
 import com.cantara.kcp.memory.scanner.AgentSessionScanner;
 import com.cantara.kcp.memory.scanner.EventLogScanner;
 import com.cantara.kcp.memory.scanner.SessionScanner;
+import com.cantara.kcp.memory.server.TcpHttpServer;
 import com.cantara.kcp.memory.store.MemoryDatabase;
 import com.cantara.kcp.memory.update.UpdateChecker;
-import com.sun.net.httpserver.HttpServer;
 
-import java.net.InetSocketAddress;
 import java.sql.SQLException;
 import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
@@ -18,14 +17,19 @@ import java.util.logging.Logger;
 
 /**
  * HTTP daemon — listens on localhost:7735.
- * <p>
- * Endpoints:
+ *
+ * <p>Uses a plain {@link TcpHttpServer} (ServerSocket + virtual threads) instead of
+ * {@code com.sun.net.httpserver.HttpServer}, avoiding the NIO selector dependency
+ * ({@code WEPollSelectorImpl} / AF_UNIX pipes) that fails inside Windows MSIX sandboxes.
+ *
+ * <p>Endpoints:
  * <pre>
  *   GET  /health          — liveness + session count
  *   GET  /search?q=...    — FTS5 full-text search
  *   GET  /sessions        — list recent sessions
  *   GET  /stats           — aggregate statistics
  *   POST /scan            — trigger incremental scan (fire-and-forget)
+ *   GET  /events/search   — tool-call event search
  * </pre>
  */
 public class KcpMemoryDaemon {
@@ -34,7 +38,7 @@ public class KcpMemoryDaemon {
     public  static final int    PORT = 7735;
 
     private final MemoryDatabase db;
-    private HttpServer server;
+    private TcpHttpServer server;
     private ScheduledExecutorService scheduler;
 
     public KcpMemoryDaemon(MemoryDatabase db) {
@@ -42,8 +46,7 @@ public class KcpMemoryDaemon {
     }
 
     public void start() throws Exception {
-        server = HttpServer.create(new InetSocketAddress("127.0.0.1", PORT), 0);
-        server.setExecutor(Executors.newVirtualThreadPerTaskExecutor());
+        server = new TcpHttpServer(PORT);
 
         server.createContext("/health",        new HealthHandler(db));
         server.createContext("/search",        new SearchHandler(db));
@@ -97,13 +100,13 @@ public class KcpMemoryDaemon {
         Runtime.getRuntime().addShutdownHook(new Thread(() -> {
             LOG.info("Shutting down kcp-memory daemon...");
             scheduler.shutdownNow();
-            server.stop(2);
+            server.stop();
             try { db.close(); } catch (SQLException ignored) {}
         }));
     }
 
     public void stop() {
         if (scheduler != null) scheduler.shutdownNow();
-        if (server    != null) server.stop(2);
+        if (server    != null) server.stop();
     }
 }

--- a/java/src/main/java/com/cantara/kcp/memory/handler/BaseHandler.java
+++ b/java/src/main/java/com/cantara/kcp/memory/handler/BaseHandler.java
@@ -1,13 +1,12 @@
 package com.cantara.kcp.memory.handler;
 
+import com.cantara.kcp.memory.server.KcpHandler;
+import com.cantara.kcp.memory.server.TcpExchange;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.SerializationFeature;
-import com.sun.net.httpserver.HttpExchange;
-import com.sun.net.httpserver.HttpHandler;
 
 import java.io.IOException;
 import java.io.InputStream;
-import java.io.OutputStream;
 import java.net.URI;
 import java.nio.charset.StandardCharsets;
 import java.util.HashMap;
@@ -16,32 +15,28 @@ import java.util.Map;
 /**
  * Shared utilities for all HTTP handlers.
  */
-public abstract class BaseHandler implements HttpHandler {
+public abstract class BaseHandler implements KcpHandler {
 
     protected static final ObjectMapper MAPPER = new ObjectMapper()
             .enable(SerializationFeature.INDENT_OUTPUT);
 
-    protected void sendJson(HttpExchange ex, int status, Object body) throws IOException {
+    protected void sendJson(TcpExchange ex, int status, Object body) throws IOException {
         byte[] bytes = MAPPER.writeValueAsBytes(body);
-        ex.getResponseHeaders().set("Content-Type", "application/json; charset=utf-8");
-        ex.sendResponseHeaders(status, bytes.length);
-        try (OutputStream os = ex.getResponseBody()) {
-            os.write(bytes);
-        }
+        ex.sendResponse(status, "application/json; charset=utf-8", bytes);
     }
 
-    protected void sendError(HttpExchange ex, int status, String message) throws IOException {
+    protected void sendError(TcpExchange ex, int status, String message) throws IOException {
         sendJson(ex, status, Map.of("error", message));
     }
 
-    protected String readBody(HttpExchange ex) throws IOException {
+    protected String readBody(TcpExchange ex) throws IOException {
         try (InputStream is = ex.getRequestBody()) {
             return new String(is.readAllBytes(), StandardCharsets.UTF_8);
         }
     }
 
     /** Parse query string into a map. */
-    protected Map<String, String> queryParams(HttpExchange ex) {
+    protected Map<String, String> queryParams(TcpExchange ex) {
         URI uri = ex.getRequestURI();
         String query = uri.getQuery();
         Map<String, String> params = new HashMap<>();

--- a/java/src/main/java/com/cantara/kcp/memory/handler/EventsHandler.java
+++ b/java/src/main/java/com/cantara/kcp/memory/handler/EventsHandler.java
@@ -2,48 +2,45 @@ package com.cantara.kcp.memory.handler;
 
 import com.cantara.kcp.memory.model.ToolEvent;
 import com.cantara.kcp.memory.scanner.EventLogScanner;
+import com.cantara.kcp.memory.server.TcpExchange;
 import com.cantara.kcp.memory.store.EventStore;
 import com.cantara.kcp.memory.store.MemoryDatabase;
-import com.fasterxml.jackson.databind.ObjectMapper;
-import com.sun.net.httpserver.HttpExchange;
-import com.sun.net.httpserver.HttpHandler;
 
 import java.io.IOException;
-import java.io.OutputStream;
 import java.net.URLDecoder;
 import java.nio.charset.StandardCharsets;
 import java.sql.SQLException;
+import java.util.Collections;
 import java.util.List;
 
 /**
  * GET /events/search?q=...&limit=20
- * <p>
- * Full-text search over tool-call events ingested from ~/.kcp/events.jsonl.
+ *
+ * <p>Full-text search over tool-call events ingested from ~/.kcp/events.jsonl.
  * Triggers an incremental EventLogScanner pass before searching so that
  * the most recently logged events are always visible.
  */
-public class EventsHandler implements HttpHandler {
+public class EventsHandler extends BaseHandler {
 
     private final MemoryDatabase db;
-    private final ObjectMapper   mapper = new ObjectMapper();
 
     public EventsHandler(MemoryDatabase db) {
         this.db = db;
     }
 
     @Override
-    public void handle(HttpExchange exchange) throws IOException {
-        if (!"GET".equals(exchange.getRequestMethod())) {
-            exchange.sendResponseHeaders(405, -1);
-            exchange.close();
+    public void handle(TcpExchange ex) throws IOException {
+        if (!"GET".equals(ex.getRequestMethod())) {
+            sendError(ex, 405, "Method not allowed");
             return;
         }
 
-        String q     = param(exchange.getRequestURI().getQuery(), "q");
-        int    limit = intParam(exchange.getRequestURI().getQuery(), "limit", 20);
+        String query = ex.getRequestURI().getQuery();
+        String q     = param(query, "q");
+        int    limit = intParam(query, "limit", 20);
 
         if (q == null || q.isBlank()) {
-            sendJson(exchange, mapper.writeValueAsBytes(List.of()));
+            sendJson(ex, 200, Collections.emptyList());
             return;
         }
 
@@ -52,20 +49,10 @@ public class EventsHandler implements HttpHandler {
 
         try {
             List<ToolEvent> results = new EventStore(db).search(q, limit);
-            sendJson(exchange, mapper.writeValueAsBytes(results));
+            sendJson(ex, 200, results);
         } catch (SQLException e) {
-            exchange.sendResponseHeaders(500, -1);
-            exchange.close();
+            sendError(ex, 500, "Search failed: " + e.getMessage());
         }
-    }
-
-    private void sendJson(HttpExchange exchange, byte[] bytes) throws IOException {
-        exchange.getResponseHeaders().add("Content-Type", "application/json");
-        exchange.sendResponseHeaders(200, bytes.length);
-        try (OutputStream os = exchange.getResponseBody()) {
-            os.write(bytes);
-        }
-        exchange.close();
     }
 
     private String param(String query, String name) {

--- a/java/src/main/java/com/cantara/kcp/memory/handler/HealthHandler.java
+++ b/java/src/main/java/com/cantara/kcp/memory/handler/HealthHandler.java
@@ -2,8 +2,8 @@ package com.cantara.kcp.memory.handler;
 
 import com.cantara.kcp.memory.mcp.McpServer;
 import com.cantara.kcp.memory.store.MemoryDatabase;
+import com.cantara.kcp.memory.server.TcpExchange;
 import com.cantara.kcp.memory.store.SessionStore;
-import com.sun.net.httpserver.HttpExchange;
 
 import java.io.IOException;
 import java.util.Map;
@@ -20,7 +20,7 @@ public class HealthHandler extends BaseHandler {
     }
 
     @Override
-    public void handle(HttpExchange ex) throws IOException {
+    public void handle(TcpExchange ex) throws IOException {
         if (!"GET".equalsIgnoreCase(ex.getRequestMethod())) {
             sendError(ex, 405, "Method not allowed");
             return;

--- a/java/src/main/java/com/cantara/kcp/memory/handler/ListHandler.java
+++ b/java/src/main/java/com/cantara/kcp/memory/handler/ListHandler.java
@@ -2,8 +2,8 @@ package com.cantara.kcp.memory.handler;
 
 import com.cantara.kcp.memory.model.SearchResult;
 import com.cantara.kcp.memory.store.MemoryDatabase;
+import com.cantara.kcp.memory.server.TcpExchange;
 import com.cantara.kcp.memory.store.SessionStore;
-import com.sun.net.httpserver.HttpExchange;
 
 import java.io.IOException;
 import java.util.List;
@@ -21,7 +21,7 @@ public class ListHandler extends BaseHandler {
     }
 
     @Override
-    public void handle(HttpExchange ex) throws IOException {
+    public void handle(TcpExchange ex) throws IOException {
         if (!"GET".equalsIgnoreCase(ex.getRequestMethod())) {
             sendError(ex, 405, "Method not allowed");
             return;

--- a/java/src/main/java/com/cantara/kcp/memory/handler/ScanHandler.java
+++ b/java/src/main/java/com/cantara/kcp/memory/handler/ScanHandler.java
@@ -1,8 +1,8 @@
 package com.cantara.kcp.memory.handler;
 
 import com.cantara.kcp.memory.scanner.SessionScanner;
+import com.cantara.kcp.memory.server.TcpExchange;
 import com.cantara.kcp.memory.store.MemoryDatabase;
-import com.sun.net.httpserver.HttpExchange;
 
 import java.io.IOException;
 import java.util.LinkedHashMap;
@@ -21,7 +21,7 @@ public class ScanHandler extends BaseHandler {
     }
 
     @Override
-    public void handle(HttpExchange ex) throws IOException {
+    public void handle(TcpExchange ex) throws IOException {
         if (!"POST".equalsIgnoreCase(ex.getRequestMethod())) {
             sendError(ex, 405, "Method not allowed");
             return;

--- a/java/src/main/java/com/cantara/kcp/memory/handler/SearchHandler.java
+++ b/java/src/main/java/com/cantara/kcp/memory/handler/SearchHandler.java
@@ -2,8 +2,8 @@ package com.cantara.kcp.memory.handler;
 
 import com.cantara.kcp.memory.model.SearchResult;
 import com.cantara.kcp.memory.store.MemoryDatabase;
+import com.cantara.kcp.memory.server.TcpExchange;
 import com.cantara.kcp.memory.store.SessionStore;
-import com.sun.net.httpserver.HttpExchange;
 
 import java.io.IOException;
 import java.util.List;
@@ -21,7 +21,7 @@ public class SearchHandler extends BaseHandler {
     }
 
     @Override
-    public void handle(HttpExchange ex) throws IOException {
+    public void handle(TcpExchange ex) throws IOException {
         if (!"GET".equalsIgnoreCase(ex.getRequestMethod())) {
             sendError(ex, 405, "Method not allowed");
             return;

--- a/java/src/main/java/com/cantara/kcp/memory/handler/StatsHandler.java
+++ b/java/src/main/java/com/cantara/kcp/memory/handler/StatsHandler.java
@@ -2,8 +2,8 @@ package com.cantara.kcp.memory.handler;
 
 import com.cantara.kcp.memory.store.MemoryDatabase;
 import com.cantara.kcp.memory.store.SessionStore;
+import com.cantara.kcp.memory.server.TcpExchange;
 import com.cantara.kcp.memory.store.ToolUsageStore;
-import com.sun.net.httpserver.HttpExchange;
 
 import java.io.IOException;
 import java.nio.file.Files;
@@ -35,7 +35,7 @@ public class StatsHandler extends BaseHandler {
     }
 
     @Override
-    public void handle(HttpExchange ex) throws IOException {
+    public void handle(TcpExchange ex) throws IOException {
         if (!"GET".equalsIgnoreCase(ex.getRequestMethod())) {
             sendError(ex, 405, "Method not allowed");
             return;

--- a/java/src/main/java/com/cantara/kcp/memory/server/KcpHandler.java
+++ b/java/src/main/java/com/cantara/kcp/memory/server/KcpHandler.java
@@ -1,0 +1,12 @@
+package com.cantara.kcp.memory.server;
+
+import java.io.IOException;
+
+/**
+ * Minimal HTTP handler interface — replaces {@code com.sun.net.httpserver.HttpHandler}.
+ * Avoids the NIO-selector dependency that blocks Java 21 inside MSIX sandboxes.
+ */
+@FunctionalInterface
+public interface KcpHandler {
+    void handle(TcpExchange exchange) throws IOException;
+}

--- a/java/src/main/java/com/cantara/kcp/memory/server/TcpExchange.java
+++ b/java/src/main/java/com/cantara/kcp/memory/server/TcpExchange.java
@@ -1,0 +1,132 @@
+package com.cantara.kcp.memory.server;
+
+import java.io.*;
+import java.net.Socket;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Thin HTTP/1.1 request+response wrapper over a plain {@link Socket}.
+ *
+ * <p>Replaces {@code com.sun.net.httpserver.HttpExchange}. Uses blocking TCP I/O so
+ * that Java 21's NIO selectors (and the AF_UNIX pipes they require on Windows) are
+ * never touched. Safe inside MSIX sandboxes where AF_UNIX is unavailable.
+ *
+ * <p>Usage: construct from an accepted {@link Socket}, call handler logic, then
+ * {@link #close()} (or use try-with-resources).
+ */
+public class TcpExchange implements Closeable {
+
+    private static final Map<Integer, String> STATUS_TEXT = Map.of(
+            200, "OK",
+            202, "Accepted",
+            400, "Bad Request",
+            404, "Not Found",
+            405, "Method Not Allowed",
+            500, "Internal Server Error"
+    );
+
+    private final Socket socket;
+    private final OutputStream out;
+    private final String method;
+    private final URI uri;
+    private final byte[] requestBody;
+
+    public TcpExchange(Socket socket) throws IOException {
+        this.socket = socket;
+        this.out = socket.getOutputStream();
+
+        InputStream in = socket.getInputStream();
+
+        // --- Parse request line and headers byte-by-byte (safe with binary body) ---
+        String requestLine = readLine(in);
+        if (requestLine == null || requestLine.isBlank()) {
+            throw new IOException("Empty HTTP request");
+        }
+
+        String[] parts = requestLine.split(" ", 3);
+        this.method = parts.length > 0 ? parts[0] : "GET";
+        this.uri = URI.create(parts.length > 1 ? parts[1] : "/");
+
+        // Parse headers
+        int contentLength = 0;
+        String line;
+        while ((line = readLine(in)) != null && !line.isEmpty()) {
+            int colon = line.indexOf(':');
+            if (colon > 0) {
+                String key = line.substring(0, colon).trim().toLowerCase();
+                String value = line.substring(colon + 1).trim();
+                if ("content-length".equals(key)) {
+                    try { contentLength = Integer.parseInt(value); }
+                    catch (NumberFormatException ignored) {}
+                }
+            }
+        }
+
+        // Read body (exact bytes — no charset conversion)
+        if (contentLength > 0) {
+            requestBody = new byte[contentLength];
+            int totalRead = 0;
+            while (totalRead < contentLength) {
+                int read = in.read(requestBody, totalRead, contentLength - totalRead);
+                if (read < 0) break;
+                totalRead += read;
+            }
+        } else {
+            requestBody = new byte[0];
+        }
+    }
+
+    public String getRequestMethod() { return method; }
+
+    public URI getRequestURI() { return uri; }
+
+    public InputStream getRequestBody() { return new ByteArrayInputStream(requestBody); }
+
+    /**
+     * Send a complete HTTP response and flush.
+     *
+     * @param statusCode   HTTP status (200, 400, 404, 405, 500, …)
+     * @param contentType  value for Content-Type header
+     * @param body         response body bytes (may be empty)
+     */
+    public void sendResponse(int statusCode, String contentType, byte[] body) throws IOException {
+        String reason = STATUS_TEXT.getOrDefault(statusCode, "Unknown");
+        StringBuilder sb = new StringBuilder();
+        sb.append("HTTP/1.1 ").append(statusCode).append(' ').append(reason).append("\r\n");
+        sb.append("Content-Type: ").append(contentType).append("\r\n");
+        sb.append("Content-Length: ").append(body.length).append("\r\n");
+        sb.append("Connection: close\r\n");
+        sb.append("\r\n");
+        out.write(sb.toString().getBytes(StandardCharsets.UTF_8));
+        if (body.length > 0) out.write(body);
+        out.flush();
+    }
+
+    @Override
+    public void close() {
+        try { socket.close(); } catch (IOException ignored) {}
+    }
+
+    // --- helpers ---
+
+    /** Read one CRLF-terminated line from the stream (returns null on EOF). */
+    private static String readLine(InputStream in) throws IOException {
+        ByteArrayOutputStream buf = new ByteArrayOutputStream();
+        int prev = -1;
+        int b;
+        while ((b = in.read()) != -1) {
+            if (b == '\n' && prev == '\r') {
+                byte[] bytes = buf.toByteArray();
+                // strip trailing \r
+                return new String(bytes, 0, bytes.length - 1, StandardCharsets.US_ASCII);
+            }
+            buf.write(b);
+            prev = b;
+        }
+        String s = buf.toString(StandardCharsets.US_ASCII);
+        return s.isEmpty() ? null : s;
+    }
+}

--- a/java/src/main/java/com/cantara/kcp/memory/server/TcpHttpServer.java
+++ b/java/src/main/java/com/cantara/kcp/memory/server/TcpHttpServer.java
@@ -1,0 +1,97 @@
+package com.cantara.kcp.memory.server;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.logging.Logger;
+
+/**
+ * Minimal HTTP/1.1 server backed by a plain {@link ServerSocket}.
+ *
+ * <p>Drop-in replacement for {@code com.sun.net.httpserver.HttpServer} that avoids
+ * Java 21's NIO selector machinery ({@code WEPollSelectorImpl} / {@code PipeImpl})
+ * which requires AF_UNIX sockets — unavailable inside Windows MSIX sandboxes.
+ *
+ * <p>Each accepted connection is dispatched to a virtual thread. Handlers are matched
+ * by path prefix (longest wins, exact match preferred).
+ */
+public class TcpHttpServer {
+
+    private static final Logger LOG = Logger.getLogger(TcpHttpServer.class.getName());
+
+    private final int port;
+    private final Map<String, KcpHandler> handlers = new LinkedHashMap<>();
+
+    private volatile boolean running = false;
+    private ServerSocket serverSocket;
+
+    public TcpHttpServer(int port) {
+        this.port = port;
+    }
+
+    /** Register a handler for the given path prefix (exact match preferred at dispatch time). */
+    public void createContext(String path, KcpHandler handler) {
+        handlers.put(path, handler);
+    }
+
+    /** Bind the socket and start the acceptor virtual thread. */
+    public void start() throws IOException {
+        serverSocket = new ServerSocket();
+        serverSocket.setReuseAddress(true);
+        serverSocket.bind(new InetSocketAddress("127.0.0.1", port));
+        running = true;
+
+        Thread.ofVirtual().name("kcp-tcp-acceptor").start(() -> {
+            while (running) {
+                try {
+                    Socket conn = serverSocket.accept();
+                    Thread.ofVirtual().start(() -> dispatch(conn));
+                } catch (IOException e) {
+                    if (running) LOG.fine("Acceptor error: " + e.getMessage());
+                }
+            }
+        });
+    }
+
+    /** Stop accepting new connections. In-flight requests are not interrupted. */
+    public void stop() {
+        running = false;
+        try { if (serverSocket != null) serverSocket.close(); } catch (IOException ignored) {}
+    }
+
+    // --- internal ---
+
+    private void dispatch(Socket conn) {
+        try (TcpExchange ex = new TcpExchange(conn)) {
+            KcpHandler handler = resolve(ex.getRequestURI().getPath());
+            if (handler != null) {
+                handler.handle(ex);
+            } else {
+                ex.sendResponse(404, "application/json; charset=utf-8",
+                        "{\"error\":\"Not found\"}".getBytes());
+            }
+        } catch (IOException e) {
+            LOG.fine("Request handling error: " + e.getMessage());
+        }
+    }
+
+    /** Exact match first, then longest-prefix match. */
+    private KcpHandler resolve(String path) {
+        KcpHandler exact = handlers.get(path);
+        if (exact != null) return exact;
+
+        KcpHandler best = null;
+        int bestLen = -1;
+        for (Map.Entry<String, KcpHandler> entry : handlers.entrySet()) {
+            String prefix = entry.getKey();
+            if (path.startsWith(prefix) && prefix.length() > bestLen) {
+                best = entry.getValue();
+                bestLen = prefix.length();
+            }
+        }
+        return best;
+    }
+}

